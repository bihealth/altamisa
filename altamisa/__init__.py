# -*- coding: utf-8 -*-
"""Top-level package for AltamISA."""

__author__ = 'Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>'
# TODO: add version with versioneer
