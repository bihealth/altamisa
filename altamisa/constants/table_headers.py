# -*- coding: utf-8 -*-
"""
Constants of valid column headers for study and assay table parsing and writing,
as specified in the `ISA-Tab study and assay specifications
<https://isa-specs.readthedocs.io/en/latest/isatab.html#study-and-assay-files>`_.
"""


__author__ = (
    "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>, "
    "Mathias Kuhring <mathias.kuhring@bihealth.de>"
)


# The following constants are used for further qualifying column types of
# material and process nodes as well as annotations.


# Material types (nodes)
EXTRACT_NAME = "Extract Name"  #:
LABELED_EXTRACT_NAME = "Labeled Extract Name"  #:
LIBRARY_NAME = "Library Name"  #: Special material (see :ref:`special_extensions`)
SAMPLE_NAME = "Sample Name"  #:
SOURCE_NAME = "Source Name"  #:


# Data types (nodes)
ARRAY_DATA_FILE = "Array Data File"  #:
ARRAY_DATA_MATRIX_FILE = "Array Data Matrix File"  #:
ARRAY_DESIGN_FILE = "Array Design File"  #:
DERIVED_ARRAY_DATA_FILE = "Derived Array Data File"  #:
DERIVED_ARRAY_DATA_MATRIX_FILE = "Derived Array Data Matrix File"  #:
DERIVED_DATA_FILE = "Derived Data File"  #:
DERIVED_SPECTRAL_DATA_FILE = "Derived Spectral Data File"  #:
IMAGE_FILE = "Image File"  #:
METABOLITE_ASSIGNMENT_FILE = "Metabolite Assignment File"  #:
PEPTIDE_ASSIGNMENT_FILE = "Peptide Assignment File"  #:
POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE = (
    "Post Translational Modification Assignment File"
)  #:
PROTEIN_ASSIGNMENT_FILE = "Protein Assignment File"  #:
RAW_DATA_FILE = "Raw Data File"  #:
RAW_SPECTRAL_DATA_FILE = "Raw Spectral Data File"  #:
SPOT_PICKING_FILE = "Spot Picking File"  #:


# Assay types (nodes)
ASSAY_NAME = "Assay Name"  #:
DATA_TRANSFORMATION_NAME = "Data Transformation Name"  #:
GEL_ELECTROPHORESIS_ASSAY_NAME = "Gel Electrophoresis Assay Name"  #:
HYBRIDIZATION_ASSAY_NAME = "Hybridization Assay Name"  #:
MS_ASSAY_NAME = "MS Assay Name"  #:
NORMALIZATION_NAME = "Normalization Name"  #:
PROTOCOL_REF = "Protocol REF"  #:
SCAN_NAME = "Scan Name"  #:


# Other simple types (standard annotations)
DATE = "Date"  #:
LABEL = "Label"  #:
MATERIAL_TYPE = "Material Type"  #:
PERFORMER = "Performer"  #:
TERM_SOURCE_REF = "Term Source REF"  #:
TERM_ACCESSION_NUMBER = "Term Accession Number"  #:
UNIT = "Unit"  #:


# Other simple types (special case annotations)
ARRAY_DESIGN_REF = "Array Design REF"  #:
FIRST_DIMENSION = "First Dimension"  #:
SECOND_DIMENSION = "Second Dimension"  #:


# Labeled types (annotations)
CHARACTERISTICS = "Characteristics"  #:
COMMENT = "Comment"  #:
FACTOR_VALUE = "Factor Value"  #:
PARAMETER_VALUE = "Parameter Value"  #:


#: Collected header keys indicating a data file.
DATA_FILE_HEADERS = (
    ARRAY_DATA_FILE,
    ARRAY_DATA_MATRIX_FILE,
    ARRAY_DESIGN_FILE,
    DERIVED_ARRAY_DATA_FILE,
    DERIVED_ARRAY_DATA_MATRIX_FILE,
    DERIVED_DATA_FILE,
    DERIVED_SPECTRAL_DATA_FILE,
    IMAGE_FILE,
    METABOLITE_ASSIGNMENT_FILE,
    PEPTIDE_ASSIGNMENT_FILE,
    POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE,
    PROTEIN_ASSIGNMENT_FILE,
    RAW_DATA_FILE,
    RAW_SPECTRAL_DATA_FILE,
    SPOT_PICKING_FILE,
)


#: Collected header keys indicating a material/data name.
MATERIAL_NAME_HEADERS = (
    # Material
    EXTRACT_NAME,
    LABELED_EXTRACT_NAME,
    LIBRARY_NAME,
    SAMPLE_NAME,
    SOURCE_NAME,
    # Data
    *DATA_FILE_HEADERS,
)


#: Collected header keys indicating a process name.
PROCESS_NAME_HEADERS = (
    ASSAY_NAME,
    DATA_TRANSFORMATION_NAME,
    GEL_ELECTROPHORESIS_ASSAY_NAME,
    HYBRIDIZATION_ASSAY_NAME,
    MS_ASSAY_NAME,
    NORMALIZATION_NAME,
    SCAN_NAME,
    # PROTOCOL_REF,  # is not a name header!
)
