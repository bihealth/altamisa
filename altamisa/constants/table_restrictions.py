# -*- coding: utf-8 -*-
"""Constants used for validation of data and annotation restrictions in study and assay tables."""


__author__ = (
    "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>, "
    "Mathias Kuhring <mathias.kuhring@bihealth.de>"
)


from . import table_headers


# Assay measurement types (only the once needed for special validations)
PROTEIN_EXPRESSION_PROFILING = "protein expression profiling"  #:
PROTEIN_IDENTIFICATION = "protein identification"  #:
METABOLITE_PROFILING = "metabolite profiling"  #:


# Assay technology types (only the once needed for special validations)
DNA_MICROARRAY = "dna microarray"  #:
GEL_ELECTROPHORESIS = "gel electrophoresis"  #:
PROTEIN_MICROARRAY = "protein microarray"  #:
NUCLEOTIDE_SEQUENCING = "nucleotide sequencing"  #:
MASS_SPECTROMETRY = "mass spectrometry"  #:


# Protocol types (only the once needed for special validations)
PT_DATA_COLLECTION = "data collection"  #:
PT_DATA_NORMALIZATION = "data normalization"  #:
PT_DATA_TRANSFORMATION = "data transformation"  #:
PT_ELECTROPHORESIS = "electrophoresis"  #:
PT_HYBRIDIZATION = "hybridization"  #:
PT_MASS_SPECTROMETRY = "mass spectrometry"  #:
PT_NUCLEIC_ACID_HYBRIDIZATION = "nucleic acid hybridization"  #:


# Material restrictions ----------------------------------------------------------------------------


#: Materials restricted to certain assay technologies
RESTRICTED_MATERIALS_ATECH = {
    # nucleotide sequencing
    table_headers.LIBRARY_NAME: {NUCLEOTIDE_SEQUENCING}
}


#: Materials restricted to certain assay measurements
RESTRICTED_MATERIALS_AMEAS = {}


# Data file restrictions ---------------------------------------------------------------------------


#: Data files restricted to certain assay technologies
RESTRICTED_FILES_ATECH = {
    # microarray
    table_headers.ARRAY_DESIGN_FILE: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    table_headers.ARRAY_DATA_FILE: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    table_headers.ARRAY_DATA_MATRIX_FILE: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    table_headers.DERIVED_ARRAY_DATA_FILE: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    table_headers.DERIVED_ARRAY_DATA_MATRIX_FILE: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    # gel eletrophoresis
    table_headers.SPOT_PICKING_FILE: {GEL_ELECTROPHORESIS},
    # mass spectrometry
    table_headers.DERIVED_SPECTRAL_DATA_FILE: {MASS_SPECTROMETRY},
    table_headers.RAW_SPECTRAL_DATA_FILE: {MASS_SPECTROMETRY},
    table_headers.PEPTIDE_ASSIGNMENT_FILE: {MASS_SPECTROMETRY},
    table_headers.POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE: {MASS_SPECTROMETRY},
    table_headers.PROTEIN_ASSIGNMENT_FILE: {MASS_SPECTROMETRY},
    table_headers.METABOLITE_ASSIGNMENT_FILE: {MASS_SPECTROMETRY},
}


#: Data files restricted to certain assay measurements
RESTRICTED_FILES_AMEAS = {
    # proteomics
    table_headers.PEPTIDE_ASSIGNMENT_FILE: {PROTEIN_EXPRESSION_PROFILING, PROTEIN_IDENTIFICATION},
    table_headers.POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE: {
        PROTEIN_EXPRESSION_PROFILING,
        PROTEIN_IDENTIFICATION,
    },
    table_headers.PROTEIN_ASSIGNMENT_FILE: {PROTEIN_EXPRESSION_PROFILING, PROTEIN_IDENTIFICATION},
    # metabolomics
    table_headers.METABOLITE_ASSIGNMENT_FILE: {METABOLITE_PROFILING},
}


# Protocol restrictions ----------------------------------------------------------------------------


#: Protocol names restricted to certain assay technologies
RESTRICTED_PROTO_NAMES_ATECH = {
    # microarray
    table_headers.SCAN_NAME: {DNA_MICROARRAY, GEL_ELECTROPHORESIS, PROTEIN_MICROARRAY},
    table_headers.HYBRIDIZATION_ASSAY_NAME: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    # gel electrophoresis
    table_headers.GEL_ELECTROPHORESIS_ASSAY_NAME: {GEL_ELECTROPHORESIS},
    # mass spectrometry
    table_headers.MS_ASSAY_NAME: {MASS_SPECTROMETRY},
}

#: Protocol names restricted by certain protocol types (ignored if ref is UNKNOWN)
RESTRICTED_PROTO_NAMES_PTYPE = {
    table_headers.DATA_TRANSFORMATION_NAME: {PT_DATA_TRANSFORMATION},
    table_headers.GEL_ELECTROPHORESIS_ASSAY_NAME: {PT_ELECTROPHORESIS},
    table_headers.HYBRIDIZATION_ASSAY_NAME: {PT_HYBRIDIZATION, PT_NUCLEIC_ACID_HYBRIDIZATION},
    table_headers.MS_ASSAY_NAME: {PT_MASS_SPECTROMETRY},
    table_headers.NORMALIZATION_NAME: {PT_DATA_NORMALIZATION},
    table_headers.SCAN_NAME: {PT_DATA_COLLECTION},
}

#: Protocol special case annotations restricted to certain assay technologies
RESTRICTED_PROTO_ANNOS_ATECH = {
    # microarray
    table_headers.ARRAY_DESIGN_REF: {DNA_MICROARRAY, PROTEIN_MICROARRAY},
    # gel electrophoresis
    table_headers.FIRST_DIMENSION: {GEL_ELECTROPHORESIS},
    table_headers.SECOND_DIMENSION: {GEL_ELECTROPHORESIS},
}

#: Protocol special case annotations restricted to certain protocol types
RESTRICTED_PROTO_ANNOS_PTYPE = {
    table_headers.ARRAY_DESIGN_REF: {PT_HYBRIDIZATION, PT_NUCLEIC_ACID_HYBRIDIZATION},
    table_headers.FIRST_DIMENSION: {PT_ELECTROPHORESIS},
    table_headers.SECOND_DIMENSION: {PT_ELECTROPHORESIS},
}
