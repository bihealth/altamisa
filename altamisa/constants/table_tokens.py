# -*- coding: utf-8 -*-
"""Token constants for study and assay table parsing and writing."""


__author__ = (
    "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>, "
    "Mathias Kuhring <mathias.kuhring@bihealth.de>"
)


#: Used for marking anonymous/unnamed Processes
TOKEN_ANONYMOUS = "Anonymous"
#: Used for marking empty/unnamed Data
TOKEN_EMPTY = "Empty"
#: Used for named Processes without protocol reference
TOKEN_UNKNOWN = "Unknown"
