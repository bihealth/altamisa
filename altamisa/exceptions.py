# -*- coding: utf-8 -*-
"""
Exceptions and Warnings used in the AltamISA library.
"""

__author__ = "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>"


class IsaException(Exception):
    """Base class for exceptions raised by Altamisa."""


class ParseIsatabException(IsaException):
    """Exception raised on problems parsing ISA-TAB."""


class WriteIsatabException(IsaException):
    """Exception raised on problems writing ISA-TAB."""


class IsaWarning(Warning):
    """Base class for warnings raised by Altamisa."""


class ParseIsatabWarning(IsaWarning):
    """Warning raised on problems parsing ISA-TAB."""


class WriteIsatabWarning(IsaWarning):
    """Warning raised on problems writing ISA-TAB."""


class IsaValidationWarning(IsaWarning):
    """Warning raised on problems validating ISA models or objects."""


class AdvisoryIsaValidationWarning(IsaValidationWarning):
    """Warning raised on uncritical problems when validating ISA models or objects."""


class ModerateIsaValidationWarning(IsaValidationWarning):
    """Warning raised on moderate problems when validating ISA models or objects."""


class CriticalIsaValidationWarning(IsaValidationWarning):
    """Warning raised on critical problems when validating ISA models or objects."""
