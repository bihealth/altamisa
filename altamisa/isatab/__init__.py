# -*- coding: utf-8 -*-
"""Module for parsing ISA-Tab files."""

# Make all models and the ``*Reader`` classes visible within this module.

from .headers import *  # noqa: F403, F401
from .models import *  # noqa: F403, F401
from .parse_assay_study import (  # noqa: F401
    AssayReader,
    AssayRowReader,
    StudyReader,
    StudyRowReader,
)
from .parse_investigation import InvestigationReader  # noqa: F401
from .validate_assay_study import AssayValidator, StudyValidator  # noqa: F401
from .validate_investigation import InvestigationValidator  # noqa: F401
from .write_assay_study import AssayWriter, StudyWriter, RefTableBuilder  # noqa: F401
from .write_investigation import InvestigationWriter  # noqa: F401
