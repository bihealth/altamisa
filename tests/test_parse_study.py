# -*- coding: utf-8 -*-
"""Tests for parsing ISA study files"""


from datetime import date
import io
import os
import pytest

from altamisa.constants import table_headers
from altamisa.exceptions import IsaWarning
from altamisa.isatab import models
from altamisa.isatab import (
    InvestigationReader,
    InvestigationValidator,
    StudyRowReader,
    StudyReader,
    StudyValidator,
)


def test_study_row_reader_minimal_study(minimal_investigation_file, minimal_study_file):
    """Use ``StudyRowReader`` to read in minimal study file."""

    # Create new row reader and check read headers
    row_reader = StudyRowReader.from_stream("S1", minimal_study_file)
    assert 3 == len(row_reader.header)

    # Read all rows in study
    rows = list(row_reader.read())

    # Check results
    assert 1 == len(rows)
    first_row = rows[0]

    assert 3 == len(first_row)

    expected = models.Material(
        "Source Name", "S1-source-0815", "0815", None, (), (), (), None, [table_headers.SOURCE_NAME]
    )
    assert expected == first_row[0]
    expected = models.Process(
        "sample collection",
        "S1-sample collection-2-1",
        None,
        None,
        None,
        None,
        (),
        (),
        None,
        None,
        None,
        [table_headers.PROTOCOL_REF],
    )
    assert expected == first_row[1]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-N1",
        "0815-N1",
        None,
        (),
        (),
        (),
        None,
        [table_headers.SAMPLE_NAME],
    )
    assert expected == first_row[2]


def test_study_reader_minimal_study(minimal_investigation_file, minimal_study_file):
    """Use ``StudyReader`` to read in minimal study file.

    Using the ``StudyReader`` instead of the ``StudyRowReader`` gives us
    ``Study`` objects instead of just the row-wise nodes.
    """
    # Load investigation (tested elsewhere)
    investigation = InvestigationReader.from_stream(minimal_investigation_file).read()
    with pytest.warns(IsaWarning) as record:
        InvestigationValidator(investigation).validate()

    # Check warnings
    assert 2 == len(record)

    # Create new row reader and check read headers
    reader = StudyReader.from_stream("S1", minimal_study_file)
    assert 3 == len(reader.header)

    # Read study
    study = reader.read()
    StudyValidator(investigation, investigation.studies[0], study).validate()

    # Check results
    assert os.path.normpath(str(study.file)).endswith(
        os.path.normpath("data/i_minimal/s_minimal.txt")
    )
    assert 3 == len(study.header)
    assert 2 == len(study.materials)
    assert 1 == len(study.processes)
    assert 2 == len(study.arcs)

    expected = models.Material(
        "Source Name", "S1-source-0815", "0815", None, (), (), (), None, [table_headers.SOURCE_NAME]
    )
    assert expected == study.materials["S1-source-0815"]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-N1",
        "0815-N1",
        None,
        (),
        (),
        (),
        None,
        [table_headers.SAMPLE_NAME],
    )
    assert expected == study.materials["S1-sample-0815-N1"]

    expected = models.Process(
        "sample collection",
        "S1-sample collection-2-1",
        None,
        None,
        None,
        None,
        (),
        (),
        None,
        None,
        None,
        [table_headers.PROTOCOL_REF],
    )
    assert expected == study.processes["S1-sample collection-2-1"]

    expected = (
        models.Arc("S1-source-0815", "S1-sample collection-2-1"),
        models.Arc("S1-sample collection-2-1", "S1-sample-0815-N1"),
    )
    assert expected == study.arcs


def test_study_row_reader_small_study(small_investigation_file, small_study_file):
    """Use ``StudyRowReader`` to read in small study file."""

    # Create new row reader and check read headers (+ string representation)
    row_reader = StudyRowReader.from_stream("S1", small_study_file)
    assert 13 == len(row_reader.header)
    rep0 = "ColumnHeader(column_type='Source Name', col_no=0, span=1)"
    rep1 = "LabeledColumnHeader(column_type='Characteristics', col_no=1, span=1, label='organism')"
    assert rep0 == repr(row_reader.header[0])
    assert rep1 == repr(row_reader.header[1])

    # Read all rows in study
    rows = list(row_reader.read())

    # Check results
    assert 5 == len(rows)
    first_row = rows[0]
    second_row = rows[1]
    third_row = rows[2]

    assert 3 == len(second_row)

    headers_source = [
        table_headers.SOURCE_NAME,
        table_headers.CHARACTERISTICS + "[organism]",
        table_headers.TERM_SOURCE_REF,
        table_headers.TERM_ACCESSION_NUMBER,
        table_headers.CHARACTERISTICS + "[age]",
        table_headers.UNIT,
        table_headers.TERM_SOURCE_REF,
        table_headers.TERM_ACCESSION_NUMBER,
    ]
    headers_collection = [
        table_headers.PROTOCOL_REF,
        table_headers.PARAMETER_VALUE + "[instrument]",
        table_headers.PERFORMER,
        table_headers.DATE,
    ]
    headers_sample = [
        table_headers.SAMPLE_NAME,
        table_headers.CHARACTERISTICS + "[status]",
        table_headers.FACTOR_VALUE + "[treatment]",
    ]

    unit = models.OntologyTermRef(
        name="day", accession="http://purl.obolibrary.org/obo/UO_0000033", ontology_name="UO"
    )

    characteristics1 = (
        models.Characteristics(
            name="organism",
            value=[
                models.OntologyTermRef(
                    name="Mus musculus",
                    accession="http://purl.bioontology.org/ontology/NCBITAXON/10090",
                    ontology_name="NCBITAXON",
                ),
                models.OntologyTermRef(
                    name="Homo sapiens",
                    accession="http://purl.bioontology.org/ontology/NCBITAXON/9606",
                    ontology_name="NCBITAXON",
                ),
            ],
            unit=None,
        ),
        models.Characteristics(name="age", value=["90"], unit=unit),
    )

    expected = models.Material(
        "Source Name",
        "S1-source-0814",
        "0814",
        None,
        characteristics1,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == first_row[0]

    characteristics2 = (
        models.Characteristics(
            name="organism",
            value=[
                models.OntologyTermRef(
                    name="Mus musculus",
                    accession="http://purl.bioontology.org/ontology/NCBITAXON/10090",
                    ontology_name="NCBITAXON",
                )
            ],
            unit=None,
        ),
        models.Characteristics(name="age", value=["90"], unit=unit),
    )

    expected = models.Material(
        "Source Name",
        "S1-source-0815",
        "0815",
        None,
        characteristics2,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == second_row[0]
    expected = models.Process(
        "sample collection",
        "S1-sample collection-9-2",
        None,
        None,
        date(2018, 2, 2),
        "John Doe",
        (models.ParameterValue("instrument", ["scalpel"], None),),
        (),
        None,
        None,
        None,
        headers_collection,
    )
    assert expected == second_row[1]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-N1",
        "0815-N1",
        None,
        (models.Characteristics("status", ["0"], None),),
        (),
        (models.FactorValue("treatment", "yes", None),),
        None,
        headers_sample,
    )
    assert expected == second_row[2]

    assert 3 == len(third_row)
    expected = models.Material(
        "Source Name",
        "S1-source-0815",
        "0815",
        None,
        characteristics2,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == third_row[0]
    expected = models.Process(
        "sample collection",
        "S1-sample collection-9-3",
        None,
        None,
        date(2018, 2, 2),
        "John Doe",
        (models.ParameterValue("instrument", ["scalpel type A", "scalpel type B"], None),),
        (),
        None,
        None,
        None,
        headers_collection,
    )
    assert expected == third_row[1]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-T1",
        "0815-T1",
        None,
        (models.Characteristics("status", ["2"], None),),
        (),
        (models.FactorValue("treatment", "", None),),
        None,
        headers_sample,
    )
    assert expected == third_row[2]


def test_study_reader_small_study(small_investigation_file, small_study_file):
    """Use ``StudyReader`` to read in small study file."""
    # Load investigation (tested elsewhere)
    with pytest.warns(IsaWarning) as record:
        investigation = InvestigationReader.from_stream(small_investigation_file).read()
        InvestigationValidator(investigation).validate()

    # Check warnings
    assert 2 == len(record)

    # Create new row reader and check read headers
    reader = StudyReader.from_stream("S1", small_study_file)
    assert 13 == len(reader.header)

    # Read study
    study = reader.read()
    StudyValidator(investigation, investigation.studies[0], study).validate()

    # Check results
    assert os.path.normpath(str(study.file)).endswith(os.path.normpath("data/i_small/s_small.txt"))
    assert 13 == len(study.header)
    assert 9 == len(study.materials)
    assert 5 == len(study.processes)
    assert 10 == len(study.arcs)

    headers_source = [
        table_headers.SOURCE_NAME,
        table_headers.CHARACTERISTICS + "[organism]",
        table_headers.TERM_SOURCE_REF,
        table_headers.TERM_ACCESSION_NUMBER,
        table_headers.CHARACTERISTICS + "[age]",
        table_headers.UNIT,
        table_headers.TERM_SOURCE_REF,
        table_headers.TERM_ACCESSION_NUMBER,
    ]
    headers_collection = [
        table_headers.PROTOCOL_REF,
        table_headers.PARAMETER_VALUE + "[instrument]",
        table_headers.PERFORMER,
        table_headers.DATE,
    ]
    headers_sample = [
        table_headers.SAMPLE_NAME,
        table_headers.CHARACTERISTICS + "[status]",
        table_headers.FACTOR_VALUE + "[treatment]",
    ]

    unit = models.OntologyTermRef(
        name="day", accession="http://purl.obolibrary.org/obo/UO_0000033", ontology_name="UO"
    )

    characteristics1 = (
        models.Characteristics(
            name="organism",
            value=[
                models.OntologyTermRef(
                    name="Mus musculus",
                    accession="http://purl.bioontology.org/ontology/" "NCBITAXON/10090",
                    ontology_name="NCBITAXON",
                )
            ],
            unit=None,
        ),
        models.Characteristics(name="age", value=["90"], unit=unit),
    )
    characteristics2 = (
        models.Characteristics(
            name="organism", value=[models.OntologyTermRef("Mus musculus", "", "")], unit=None
        ),
        models.Characteristics(name="age", value=[""], unit=unit),
    )
    characteristics3 = (
        models.Characteristics(
            name="organism", value=[models.OntologyTermRef(None, None, None)], unit=None
        ),
        models.Characteristics(name="age", value=["150"], unit=unit),
    )

    expected = models.Material(
        "Source Name",
        "S1-source-0815",
        "0815",
        None,
        characteristics1,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == study.materials["S1-source-0815"]
    expected = models.Material(
        "Source Name",
        "S1-source-0816",
        "0816",
        None,
        characteristics2,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == study.materials["S1-source-0816"]
    expected = models.Material(
        "Source Name",
        "S1-source-0817",
        "0817",
        None,
        characteristics3,
        (),
        (),
        None,
        headers_source,
    )
    assert expected == study.materials["S1-source-0817"]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-N1",
        "0815-N1",
        None,
        (models.Characteristics("status", ["0"], None),),
        (),
        (models.FactorValue("treatment", "yes", None),),
        None,
        headers_sample,
    )
    assert expected == study.materials["S1-sample-0815-N1"]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0815-T1",
        "0815-T1",
        None,
        (models.Characteristics("status", ["2"], None),),
        (),
        (models.FactorValue("treatment", "", None),),
        None,
        headers_sample,
    )
    assert expected == study.materials["S1-sample-0815-T1"]
    expected = models.Material(
        "Sample Name",
        "S1-sample-0816-T1",
        "0816-T1",
        None,
        (models.Characteristics("status", ["1"], None),),
        (),
        (models.FactorValue("treatment", "yes", None),),
        None,
        headers_sample,
    )
    assert expected == study.materials["S1-sample-0816-T1"]
    expected = models.Material(
        "Sample Name",
        "S1-Empty Sample Name-13-5",
        "",
        None,
        (models.Characteristics("status", [""], None),),
        (),
        (models.FactorValue("treatment", "", None),),
        None,
        headers_sample,
    )
    assert expected == study.materials["S1-Empty Sample Name-13-5"]

    expected = models.Process(
        "sample collection",
        "S1-sample collection-9-2",
        None,
        None,
        date(2018, 2, 2),
        "John Doe",
        (models.ParameterValue("instrument", ["scalpel"], None),),
        (),
        None,
        None,
        None,
        headers_collection,
    )
    assert expected == study.processes["S1-sample collection-9-2"]
    expected = models.Process(
        "sample collection",
        "S1-sample collection-9-3",
        None,
        None,
        date(2018, 2, 2),
        "John Doe",
        (models.ParameterValue("instrument", ["scalpel type A", "scalpel type B"], None),),
        (),
        None,
        None,
        None,
        headers_collection,
    )
    assert expected == study.processes["S1-sample collection-9-3"]
    expected = models.Process(
        "sample collection",
        "S1-sample collection-9-4",
        None,
        None,
        date(2018, 2, 2),
        "John Doe",
        (models.ParameterValue("instrument", ["scalpel"], None),),
        (),
        None,
        None,
        None,
        headers_collection,
    )
    assert expected == study.processes["S1-sample collection-9-4"]

    expected = (
        models.Arc("S1-source-0814", "S1-sample collection-9-1"),
        models.Arc("S1-sample collection-9-1", "S1-sample-0814-N1"),
        models.Arc("S1-source-0815", "S1-sample collection-9-2"),
        models.Arc("S1-sample collection-9-2", "S1-sample-0815-N1"),
        models.Arc("S1-source-0815", "S1-sample collection-9-3"),
        models.Arc("S1-sample collection-9-3", "S1-sample-0815-T1"),
        models.Arc("S1-source-0816", "S1-sample collection-9-4"),
        models.Arc("S1-sample collection-9-4", "S1-sample-0816-T1"),
        models.Arc("S1-source-0817", "S1-sample collection-9-5"),
        models.Arc("S1-sample collection-9-5", "S1-Empty Sample Name-13-5"),
    )
    assert expected == study.arcs


def test_study_reader_minimal_study_iostring(minimal_investigation_file, minimal_study_file):
    # Load investigation (tested elsewhere)
    stringio = io.StringIO(minimal_investigation_file.read())
    investigation = InvestigationReader.from_stream(stringio).read()
    with pytest.warns(IsaWarning) as record:
        InvestigationValidator(investigation).validate()

    # Check warnings
    assert 2 == len(record)

    # Create new study reader and read from StringIO with original filename indicated
    stringio = io.StringIO(minimal_study_file.read())
    reader = StudyReader.from_stream("S1", stringio, "data/i_minimal/s_minimal.txt")
    assert 3 == len(reader.header)

    # Read study
    study = reader.read()
    StudyValidator(investigation, investigation.studies[0], study).validate()

    # Check results
    assert os.path.normpath(str(study.file)).endswith(
        os.path.normpath("data/i_minimal/s_minimal.txt")
    )
    assert 3 == len(study.header)
    assert 2 == len(study.materials)
    assert 1 == len(study.processes)
    assert 2 == len(study.arcs)


def test_study_reader_minimal_study_iostring2(minimal_investigation_file, minimal_study_file):
    # Load investigation (tested elsewhere)
    stringio = io.StringIO(minimal_investigation_file.read())
    investigation = InvestigationReader.from_stream(stringio).read()
    with pytest.warns(IsaWarning) as record:
        InvestigationValidator(investigation).validate()

    # Check warnings
    assert 2 == len(record)

    # Create new study reader and read from StringIO with no filename indicated
    stringio = io.StringIO(minimal_study_file.read())
    reader = StudyReader.from_stream("S1", stringio)
    assert 3 == len(reader.header)

    # Read study
    study = reader.read()
    StudyValidator(investigation, investigation.studies[0], study).validate()

    # Check results
    assert str(study.file) == "<no file>"
    assert 3 == len(study.header)
    assert 2 == len(study.materials)
    assert 1 == len(study.processes)
    assert 2 == len(study.arcs)
