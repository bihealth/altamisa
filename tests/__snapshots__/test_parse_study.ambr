# serializer version: 1
# name: test_study_reader_minimal_study
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	i_minimal.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_reader_minimal_study_iostring
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	<no file>
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_reader_minimal_study_iostring2
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	<no file>
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_reader_small_study
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_small
      	Title:	Small Investigation
      	Path:	i_small.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_small.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_reader_small_study.1
  list([
    "Found annotated material/file without name: Material(type='Sample Name', unique_name='S1-Empty Sample Name-13-5', name='', extract_label=None, characteristics=(Characteristics(name='status', value=['1'], unit=None),), comments=(), factor_values=(FactorValue(name='treatment', value=[''], unit=None),), material_type=None, headers=['Sample Name', 'Characteristics[status]', 'Factor Value[treatment]'])",
  ])
# ---
